Add Range[Inclusive]::is_empty

scottmcm opened the pull request:
During https://github.com/rust-lang/rfcs/pull/1980, it was discussed that figuring out whether a range is empty was subtle, and thus there should be a clear and obvious way to do it.  It can't just be ExactSizeIterator::is_empty (also unstable) because not all ranges are ExactSize -- such as `Range<i64>` and `RangeInclusive<usize>`.

Things to ponder:
- Unless this is stabilized first, this makes stabilizing ExactSizeIterator::is_empty more icky, since this hides that.
- This is only on `Range` and `RangeInclusive`, as those are the only ones where it's interesting.  But one could argue that it should be on more for consistency, or on RangeArgument instead.
- The bound on this is PartialOrd, since that works ok (see tests for float examples) and is consistent with `contains`.  But ranges like `NAN..=NAN`_are_ kinda weird.
- [x] ~~There's not a real issue number on this yet~~

rust-highfive commented:
r? @shepmaster

(rust_highfive has picked a reviewer for you, use r? to override)

shepmaster commented on src/libcore/iter/traits.rs:
```diff
@@ -706,7 +706,7 @@ pub trait ExactSizeIterator: Iterator {
     /// #![feature(exact_size_is_empty)]
     ///
-    /// let mut one_element = 0..1;
+    /// let mut one_element = std::iter::once(0);
```
Why is this related to this PR?

scottmcm replied:
Because using `0..1` means the code wasn't actually using `ExactSizeIterator`; the inherent method was getting picked up instead.  For normal code that's not a problem--they do exactly the same thing--but I figured the doctest should actually use the method it's documenting.

shepmaster commented on src/libcore/tests/iter.rs:
```diff
@@ -1427,9 +1427,9 @@ fn test_range_inclusive_nth() {
+    assert_eq!(ExactSizeIterator::is_empty(&r), false);
     assert_eq!(r.nth(10), None);
-    assert_eq!(r.is_empty(), true);
+    assert_eq!(ExactSizeIterator::is_empty(&r), true);
```
Is this going to be a concern for end users?

shepmaster replied:
> ExactSizeIterator::is_empty (also unstable)

ah, nvm

scottmcm replied:
`Range::is_empty` is strictly more general, so it's not a problem in a world where everything is stable.  It may be a slight hiccup in nightly, but I would expect it to be solved in the wild by adding the new feature gate, rather than the annoying UFCS.  Here I just changed it to the other method because I didn't want to change what the test was testing.  (It _would_ be a problem if we wanted to stabilize `ExactSizeIterator::is_empty` first, as mentioned in the OP, but I feel like this should be less controversial than the new trait method, especially since there are a bunch of iterators that aren't `ExactSize` that could provide an efficient `is_empty`.)

shepmaster commented:
The code seems reasonable to me, but since @kennytm has been thinking about ranges lately...

r? @kennytm

shepmaster commented on src/libcore/tests/ops.rs:
```diff
@@ -68,3 +68,27 @@ fn test_range_inclusive() {
+    assert!( (EPSILON ..= NAN).is_empty());
+    assert!( (NAN ..= EPSILON).is_empty());
+    assert!( (NAN ..= NAN).is_empty());
+}
```
WDYT about a test that calls `next` on a `Range{Inclusive}` and makes sure it transitions from not  empty to empty?

scottmcm replied:
Good thought; I'll change all of these to look for is_empty instead of `1..=0`: https://github.com/rust-lang/rust/blob/master/src/libcore/tests/iter.rs#L1433

scottmcm commented on src/libcore/tests/iter.rs:
```diff
@@ -1428,9 +1470,10 @@ fn test_range_inclusive_nth() {
     assert_eq!(r.nth(10), None);
     assert_eq!(r.is_empty(), true);
-    assert_eq!(r, 1..=0);  // We may not want to document/promise this detail
+    assert_eq!(ExactSizeIterator::is_empty(&r), true);
```
Changed RangeInclusive exhaustion tests to check `is_empty`, updated the docs to unspecify the post-iteration value of a RangeInclusive, and added some more Range tests using its `is_empty`.

kennytm commented:
Since you're defining `is_empty()` to floating point types as well, perhaps spell out in the documentation that `a..b` is not empty when `a < b`.

LGTM other than this documentation bit, but I'd also like someone from @rust-lang/libs to sign-off.

alexcrichton commented:
Looks good to me! Especially as unstable I think this can land when it's good to go

scottmcm commented:
Added the `start < end` and `start <= end` to the doc comments next to the description of what the range contains, and added a doctest for each showing the behaviour with `NAN`.

kennytm commented:
@bors r=kennytm,alexcrichton

bors commented:
:pushpin: Commit 22b0489 has been approved by `kennytm,alexcrichton`

<!-- @bors r=kennytm,alexcrichton 22b0489f80dae5242f19c4ce892b50d3685dbf82 -->
